contracts/src/AlgebraMultiQuoter.sol

```diff
@@ -0,0 +1,477 @@
+                if (step.initialized) {
+                    (, int128 liquidityNet, , , , , , ) = pool.ticks(step.tickNext);
+
+                    // if we're moving leftward, we interpret liquidityNet as the opposite sign
+                    // safe because liquidityNet cannot be type(int128).min
+                    if (zeroForOne) liquidityNet = -liquidityNet;
+
+                    state.liquidity = LiquidityMath.addDelta(state.liquidity, liquidityNet);
+                }
+
+                state.tick = zeroForOne ? step.tickNext - 1 : step.tickNext;
+                state.gasAggregate += step.gasBefore - gasleft();
+            } else if (state.sqrtPriceX96 != step.sqrtPriceStartX96) {
+                // recompute unless we're on a lower tick boundary (i.e. already transitioned ticks), and haven't moved
+                state.tick = TickMath.getTickAtSqrtRatio(state.sqrtPriceX96);
+            }
+
+            if (state.amountSpecifiedRemaining == 0) {
+                (result.amounts0[state.amountsIndex], result.amounts1[state.amountsIndex]) = zeroForOne == exactInput
+                    ? (amounts[state.amountsIndex], state.amountCalculated)
+                    : (state.amountCalculated, amounts[state.amountsIndex]);
+
+                if (state.sqrtPriceX96 != step.sqrtPriceNextX96) {
+                    result.gasEstimates[state.amountsIndex] = scaleMultiswapGasEstimate(
+                        state.gasAggregate + (step.gasBefore - gasleft())
+                    );
+                } else {
+                    // we are moving to the next tick
+                    result.gasEstimates[state.amountsIndex] = scaleMultiswapGasEstimate(state.gasAggregate);
+                }
+
+                if (state.amountsIndex == amounts.length - 1) {
+                    return (result);
+                }
+
+                state.amountsIndex += 1;
+                state.amountSpecifiedRemaining = amounts[state.amountsIndex].sub(amounts[state.amountsIndex - 1]);
+            }
+        }
+
+        for (uint256 i = state.amountsIndex; i < amounts.length; ++i) {
+            (result.amounts0[i], result.amounts1[i]) = zeroForOne == exactInput
+                ? (amounts[i] - state.amountSpecifiedRemaining, state.amountCalculated)
+                : (state.amountCalculated, amounts[i] - state.amountSpecifiedRemaining);
+            result.gasEstimates[i] = scaleMultiswapGasEstimate(state.gasAggregate);
+        }
+    }
+
+    /// @notice Returns the multiswap gas estimate scaled to AlgebraPool:swap estimates
+    /// TODO: set the parameters after running a linear regression between QuoterV2 gas estimates and unscaled
+    /// gas estimates from multiswap
+    /// @param multiSwapEstimate the gas estimate from multiswap
+    /// @return swapEstimate the gas estimate equivalent for UniswapV3Pool:swap
+    function scaleMultiswapGasEstimate(uint256 multiSwapEstimate) private pure returns (uint256 swapEstimate) {
```

Comment:
i know we follow the same pattern in the univ3 quoter as well, but it seems more natural do adjust the gas estimate post sampling (where we do the default gas schedule stuff along with the rest of the DEXs). maybe consider moving this out?

Reply:
good point. Yes, now we're basically scaling the gas estimate here once and then once more post sampling.